### PR TITLE
close serial port in case of error

### PR DESCRIPTION
**Pull request by eugenkleinschmidt**
As user cannot access the inter serial instance, the driver has to close the port in case of error

Additional:
* [make main classes accessable directly by conrad_relaycard](https://github.com/stephrdev/conrad-relaycard/commit/ad47635e1d5c47a9f010172ac30825344b886c73)
* [remove black and use ruff format instead](https://github.com/stephrdev/conrad-relaycard/commit/b4094feb228cf866f233e6046190aaa25c55c153)

### CODE REVIEW

File: tox.ini [hunk 1]
```diff
@@ -19,14 +19,14 @@ commands =
...
 [testenv:check]
 deps =
-    black
     mypy
     ruff
+    setuptools
 skip_install = true
 commands =
     python setup.py check --strict --metadata
+    ruff format .
```
Comment (stephrdev) [inline tox.ini:28]: Not sure if it is a good idea to change source files in CI or is ruff format only complaining if something isn’t formatted properly?

File: tests/test_card.py [hunk 2]
```diff
@@ -26,26 +24,26 @@ def test_relaycard():
         assert rly.is_initialized is True
         assert rly.card_count == 255
 
-        mock_serial_instance.read.return_value = b"\xFD\x00\x00\xFD"
+        mock_serial_instance.read.return_value = b"\xfd\x00\x00\xfd"
```
Comment (stephrdev) [inline tests/test_card.py:27]: Why was this chance required? Why is
it now lowercased?

File: src/conrad_relaycard/card.py [hunk 3]
```diff
@@ -36,6 +36,7 @@ def _get_serial_port(self, port: str | None = None) -> serial.Serial:
             )
 
         if not self._serial_port.is_open:
+            self._serial_port.close()
```
Comment (stephrdev) [inline src/conrad_relaycard/card.py:39]: Is it safe to call close() on a already closed serial device?

**Review by stephrdev — COMMENTED** [review]
Thank you for your contribution, some questions left :)

File: tox.ini (hunk 1 above)
Comment (eugenkleinschmidt) [reply tox.ini:28]: ruff is complete drop-in for black. Idea was to decrease the dev dependencies and only use ruff for formatting and linting

File: tests/test_card.py (hunk 2 above)
Comment (eugenkleinschmidt) [reply tests/test_card.py:27]: I think this done by ruff --format 😅

File: tox.ini (hunk 1 above)
Comment (eugenkleinschmidt) [reply tox.ini]: But good point. It need the --check option

File: src/conrad_relaycard/card.py (hunk 3 above)
Comment (eugenkleinschmidt) [reply src/conrad_relaycard/card.py]: Good question. I change it to try-close. Also forgot to put the close on every raise. That was the intention from beginning

File: src/conrad_relaycard/__init__.py [hunk 4]
```diff
@@ -0,0 +1,3 @@
+from .card import RelayCard as RelayCard
```
Comment (stephrdev) [inline src/conrad_relaycard/__init__.py:1]: the `as ..` is not needed.

Comment (eugenkleinschmidt) [reply src/conrad_relaycard/__init__.py:1]: unfortunately ruff give this warning https://docs.astral.sh/ruff/rules/unused-import/

Is disable this at this lines good for you?

```python
from .card import RelayCard  # noqa: F401
from .exceptions import RelayCardError  # noqa: F401
from .state import RelayState  # noqa: F401
```

Comment (stephrdev) [reply src/conrad_relaycard/__init__.py:1]: Yes thats better. Thanks!

Comment (eugenkleinschmidt) [reply src/conrad_relaycard/__init__.py]: Done

**Review by stephrdev — COMMENTED** [review]
Last pending thing, you mentioned to add --check but added --diff :)

File: tox.ini [hunk 5]
```diff
@@ -19,14 +19,14 @@ commands =
...
 [testenv:check]
 deps =
-    black
     mypy
     ruff
+    setuptools
 skip_install = true
 commands =
     python setup.py check --strict --metadata
+    ruff format --diff .
```
[the remark this answers is not in the document]
Comment (eugenkleinschmidt) [reply tox.ini:28]: Wouldn't it be nice to see in CI what format is expected? `--check` is not provides a diff view

[the remark this answers is not in the document]
Comment (stephrdev) [reply tox.ini:28]: Then maybe combine both? Or does --diff exist with return code 1 if there is a diff?

[the remark this answers is not in the document]
Comment (eugenkleinschmidt) [reply tox.ini:28]: Yes, it returns code 1
grafik

**Comment by eugenkleinschmidt** [comment]
> Last pending thing, you mentioned to add --check but added --diff :)

Ah... yes, my fault. I tried both and found --diff better. As mention, in CI you can see with --diff what will be changed.

**Comment by eugenkleinschmidt** [comment]
Can you create a new version on pypi? I would very appreciate it

**Comment by eugenkleinschmidt** [comment]
Many thanks 🎉 Hope it's ok when I write it here. One suggestion: I would also be nice to see the version tags here in github and also a link to pypi. Maybe in "About" details.

**Comment by stephrdev** [comment]
You are right. I’ll improve the release process :)

**Pull request merged**